firedrake/preconditioners/pmg.py

```diff
@@ -95,124 +128,147 @@ def initialize(self, pc):
-
-            cV_ = cV
-            for index in bc._indices:
-                cV_ = cV_.sub(index)
-
-            cbcs.append(firedrake.DirichletBC(cV_, firedrake.zero(cV_.shape),
-                                              bc.sub_domain,
-                                              method=bc.method))
-
-        fcp = fctx._problem.form_compiler_parameters
-        cproblem = firedrake.NonlinearVariationalProblem(cF, cu, cbcs, cJ,
-                                                         Jp=cJp,
+        cF = coarsen_form(fctx.F, Nf, Nc, replace_d)
+        cJ = coarsen_form(fctx.J, Nf, Nc, replace_d)
+        cJp = coarsen_form(fctx.Jp, Nf, Nc, replace_d)
+        fcp = coarsen_quadrature(fproblem.form_compiler_parameters, Nf, Nc)
+        cbcs = coarsen_bcs(fproblem.bcs)
+
+        # Coarsen the appctx: the user might want to provide solution-dependant expressions and forms
+        cappctx = dict(fctx.appctx)
+        for key in cappctx:
+            val = cappctx[key]
+            if isinstance(val, dict):
+                cappctx[key] = coarsen_quadrature(val, Nf, Nc)
+            elif isinstance(val, Expr):
+                cappctx[key] = replace(val, replace_d)
+            elif isinstance(val, Form):
+                cappctx[key] = coarsen_form(val, Nf, Nc, replace_d)
+
+        # Coarsen the problem and the _SNESContext
+        cproblem = firedrake.NonlinearVariationalProblem(cF, cu, bcs=cbcs, J=cJ, Jp=cJp,
                                                          form_compiler_parameters=fcp,
-                                                         is_linear=fctx._problem.is_linear)
+                                                         is_linear=fproblem.is_linear)
+
+        cctx = type(fctx)(cproblem, fctx.mat_type, fctx.pmat_type,
+                          appctx=cappctx,
+                          pre_jacobian_callback=fctx._pre_jacobian_callback,
+                          pre_function_callback=fctx._pre_function_callback,
+                          post_jacobian_callback=fctx._post_jacobian_callback,
+                          post_function_callback=fctx._post_function_callback,
+                          options_prefix=fctx.options_prefix,
+                          transfer_manager=fctx.transfer_manager)
 
-        cctx = _SNESContext(cproblem, fctx.mat_type, fctx.pmat_type,
-                            appctx=fctx.appctx,
-                            pre_jacobian_callback=fctx._pre_jacobian_callback,
-                            pre_function_callback=fctx._pre_function_callback,
-                            post_jacobian_callback=fctx._post_jacobian_callback,
-                            post_function_callback=fctx._post_function_callback,
-                            options_prefix=fctx.options_prefix,
-                            transfer_manager=fctx.transfer_manager)
+        # FIXME setting up the _fine attribute triggers gmg injection.
+        # cctx._fine = fctx
```

Comment:
This comes from the workaround in `_SNESContext.compute_operators` for a PETSc bug.